### PR TITLE
Flatten statutory sick pay tests

### PR DESCRIPTION
**Pull request by tadast**
This is up for discussion, but @BenJanecke and I think it makes it easier to change the tests. 

This PR is a result of us trying to move Q11 right after Q5 in statutory sick pay calculator, which requires a lot of tests to change. We were not able to effectively refactor tests as this task gets particularly difficult with multiple levels of nesting (16 levels of nesting was the deepest level I think). It's difficult mainly because the contexts branch out, introducing a new question requires similar "branching" at a different level, thus essentially making change harder than a complete rewrite (subjective).

Flattening tests introduces some duplication in test setup. I believe we could avoid this by using setup methods with descriptive names e.g. `setup_employee_who_noticed_on_time_about_sickness_of_4_days`. But I wanted to get feedback about this PR in general before proceeding with this idea.

Flattening the test also makes it a bit harder to write good test descriptions, but I think the ones on master need significant improvements anyway. In this particular case removing context with descriptions does not cause much harm, because reading assertions/responses is quite descriptive eg:

```ruby
        assert_current_node :employee_work_different_days?
        add_response :no
        assert_current_node :first_sick_day?
        add_response '2014-03-02'
        assert_current_node :last_sick_day?
        add_response '2014-06-02'
```

Moreover, even with some duplication the result is `259 additions and 427 deletions.` It may not be the best metric of readability, but should indicate reduction of syntactic noise in the test file.

Best reviewed in ["Split" mode](https://github.com/blog/1884-introducing-split-diffs)

Please share your thoughts :)

### CODE REVIEW

**Comment by floehopper** [comment]
How sure are you that this will help you make the changes you want to make? I think it might be good to try making the changes before we merge this.

How sure are you that we haven't lost any coverage? I'm afraid I found it really hard to follow the diffs, but that's partly because I've had my head in a couple of other things today.

**Comment by tadast** [comment]
> How sure are you that this will help you make the changes you want to make? I think it might be good to try making the changes before we merge this.

@BenJanecke do you want to rebase our WIP on this branch and try to make the changes?

> How sure are you that we haven't lost any coverage?

The overall number of assertions has grown (because of duplication), the number of tests has decreased (because not every assertion has its own test now :trollface:). It's hard to provide any scientific evidence but we paired on it and did incremental changes (moving lines outside of context, deleting empty contexts), running the tests regularly, so I'm quite confident.

**Comment by chrisroos** [comment]
Personally, I would've preferred to have seen this split into more commits to make it easier to see the changes that have been made. It'd take a long time to look at this diff and be confident that the tests are providing as much coverage as they were before.

I also wonder whether we've lost anything by removing the `context`s and `should`s. Did they contain anything that helped us understand what the test is trying to show?

**Comment by tadast** [comment]
Before:
```
CalculateStatutorySickPayTest
test_: Already getting maternity allowance Getting maternity allowance should go to result A1.  PASS (0.00s)
test_: Already getting maternity allowance Getting statutory maternity pay should go to result A1.  PASS (0.00s)
test_: Already getting maternity allowance Not getting maternity allowance employee didn't tell employer within time limit answer no answer 2 March 2014 answer 2 June 2014 answer before_payday answer irregularly answer £3000 answer 17 days answer no answer monday to friday should go to entitled_to_sick_pay outcome.  PASS (0.34s)
test_: Already getting maternity allowance Not getting maternity allowance employee didn't tell employer within time limit answer no answer 2 March 2014 answer 2 June 2014 answer before_payday answer irregularly answer £3000 answer 17 days answer no should ask which days of the week they usually work.  PASS (0.01s)
test_: Already getting maternity allowance Not getting maternity allowance employee didn't tell employer within time limit answer no answer 2 March 2014 answer 2 June 2014 answer before_payday answer irregularly answer £3000 answer 17 days should ask if employee was off sick the previous 8 weeks for 4 days.  PASS (0.00s)
test_: Already getting maternity allowance Not getting maternity allowance employee didn't tell employer within time limit answer no answer 2 March 2014 answer 2 June 2014 answer before_payday answer irregularly answer £3000 should ask how many days earnings cover.  PASS (0.00s)
test_: Already getting maternity allowance Not getting maternity allowance employee didn't tell employer within time limit answer no answer 2 March 2014 answer 2 June 2014 answer before_payday answer irregularly should ask how much they would have been paid on first payday.  PASS (0.00s)
test_: Already getting maternity allowance Not getting maternity allowance employee didn't tell employer within time limit answer no answer 2 March 2014 answer 2 June 2014 answer before_payday should ask how often employee is paid.  PASS (0.00s)
test_: Already getting maternity allowance Not getting maternity allowance employee didn't tell employer within time limit answer no answer 2 March 2014 answer 2 June 2014 should ask if employer paid 8 weeks of earnings.  PASS (0.00s)
test_: Already getting maternity allowance Not getting maternity allowance employee didn't tell employer within time limit answer no answer 2 March 2014 should ask when the last sick day was.  PASS (0.00s)
test_: Already getting maternity allowance Not getting maternity allowance employee didn't tell employer within time limit answer no should ask when the first sick day was.  PASS (0.00s)
test_: Already getting maternity allowance Not getting maternity allowance employee didn't tell employer within time limit should ask if employee works different days of the week.  PASS (0.00s)
test_: Already getting maternity allowance Not getting maternity allowance employee told employer within time limit employee works different days of the week should go to result A4.  PASS (0.00s)
test_: Already getting maternity allowance Not getting maternity allowance employee told employer within time limit employee works regular days answering first sick day answering last sick day last sick day is 3 days or more after first answer no, employee is new and fell sick before payday answer monthly answer £2000 answer 30 answer no answer weekdays should take you to result 6 without first days (no PIW).  PASS (0.01s)
test_: Already getting maternity allowance Not getting maternity allowance employee told employer within time limit employee works regular days answering first sick day answering last sick day last sick day is 3 days or more after first answer no, employee is new and fell sick before payday answer monthly answer £2000 answer 30 answer no should ask which days of the week they work.  PASS (0.00s)
test_: Already getting maternity allowance Not getting maternity allowance employee told employer within time limit employee works regular days answering first sick day answering last sick day last sick day is 3 days or more after first answer no, employee is new and fell sick before payday answer monthly answer £2000 answer 30 answer yes enter previous sickness start date answer 4 answer three days a week should take you to result A6.  PASS (0.02s)
test_: Already getting maternity allowance Not getting maternity allowance employee told employer within time limit employee works regular days answering first sick day answering last sick day last sick day is 3 days or more after first answer no, employee is new and fell sick before payday answer monthly answer £2000 answer 30 answer yes enter previous sickness start date answer 4 should ask which days they work.  PASS (0.00s)
test_: Already getting maternity allowance Not getting maternity allowance employee told employer within time limit employee works regular days answering first sick day answering last sick day last sick day is 3 days or more after first answer no, employee is new and fell sick before payday answer monthly answer £2000 answer 30 answer yes enter previous sickness start date should ask how many previous sick days were taken.  PASS (0.00s)
test_: Already getting maternity allowance Not getting maternity allowance employee told employer within time limit employee works regular days answering first sick day answering last sick day last sick day is 3 days or more after first answer no, employee is new and fell sick before payday answer monthly answer £2000 answer 30 answer yes should ask for start date of linked sickness.  PASS (0.00s)
test_: Already getting maternity allowance Not getting maternity allowance employee told employer within time limit employee works regular days answering first sick day answering last sick day last sick day is 3 days or more after first answer no, employee is new and fell sick before payday answer monthly answer £2000 answer 30 should ask abou PIW.  PASS (0.00s)
test_: Already getting maternity allowance Not getting maternity allowance employee told employer within time limit employee works regular days answering first sick day answering last sick day last sick day is 3 days or more after first answer no, employee is new and fell sick before payday answer monthly answer £2000 should ask how many days the period covers.  PASS (0.00s)
test_: Already getting maternity allowance Not getting maternity allowance employee told employer within time limit employee works regular days answering first sick day answering last sick day last sick day is 3 days or more after first answer no, employee is new and fell sick before payday answer monthly should ask how much you would have paid on their first payday.  PASS (0.00s)
test_: Already getting maternity allowance Not getting maternity allowance employee told employer within time limit employee works regular days answering first sick day answering last sick day last sick day is 3 days or more after first answer no, employee is new and fell sick before payday should ask how often you pay employees.  PASS (0.00s)
test_: Already getting maternity allowance Not getting maternity allowance employee told employer within time limit employee works regular days answering first sick day answering last sick day last sick day is 3 days or more after first answer no, paid less than 8 weeks earnings answer £3000 answer 35 days answer no answer weekdays should take you to result 6 without first days (no PIW).  PASS (0.02s)
test_: Already getting maternity allowance Not getting maternity allowance employee told employer within time limit employee works regular days answering first sick day answering last sick day last sick day is 3 days or more after first answer no, paid less than 8 weeks earnings answer £3000 answer 35 days answer no should ask which days of the week they work.  PASS (0.01s)
test_: Already getting maternity allowance Not getting maternity allowance employee told employer within time limit employee works regular days answering first sick day answering last sick day last sick day is 3 days or more after first answer no, paid less than 8 weeks earnings answer £3000 answer 35 days answer yes enter previous sickness start date answer 5 days answer weekdays should take you to result A6.  PASS (0.03s)
test_: Already getting maternity allowance Not getting maternity allowance employee told employer within time limit employee works regular days answering first sick day answering last sick day last sick day is 3 days or more after first answer no, paid less than 8 weeks earnings answer £3000 answer 35 days answer yes enter previous sickness start date answer 5 days should ask which days they work.  PASS (0.01s)
test_: Already getting maternity allowance Not getting maternity allowance employee told employer within time limit employee works regular days answering first sick day answering last sick day last sick day is 3 days or more after first answer no, paid less than 8 weeks earnings answer £3000 answer 35 days answer yes enter previous sickness start date should ask how many previous sick days were taken.  PASS (0.01s)
test_: Already getting maternity allowance Not getting maternity allowance employee told employer within time limit employee works regular days answering first sick day answering last sick day last sick day is 3 days or more after first answer no, paid less than 8 weeks earnings answer £3000 answer 35 days answer yes should ask for start date of previous sickness.  PASS (0.01s)
test_: Already getting maternity allowance Not getting maternity allowance employee told employer within time limit employee works regular days answering first sick day answering last sick day last sick day is 3 days or more after first answer no, paid less than 8 weeks earnings answer £3000 answer 35 days should ask the PIW question.  PASS (0.00s)
test_: Already getting maternity allowance Not getting maternity allowance employee told employer within time limit employee works regular days answering first sick day answering last sick day last sick day is 3 days or more after first answer no, paid less than 8 weeks earnings answer £3000 should ask how many days does this period cover.  PASS (0.00s)
test_: Already getting maternity allowance Not getting maternity allowance employee told employer within time limit employee works regular days answering first sick day answering last sick day last sick day is 3 days or more after first answer no, paid less than 8 weeks earnings should ask what total earnings before sick start date.  PASS (0.00s)
test_: Already getting maternity allowance Not getting maternity allowance employee told employer within time limit employee works regular days answering first sick day answering last sick day last sick day is 3 days or more after first answer yes, paid at least 8 weeks answer weekly enter last payday before start of sickness enter last payday before offset enter total amount paid between paydays answer no answer weekdays should take you to result 6 without first days (no PIW).  PASS (0.01s)
test_: Already getting maternity allowance Not getting maternity allowance employee told employer within time limit employee works regular days answering first sick day answering last sick day last sick day is 3 days or more after first answer yes, paid at least 8 weeks answer weekly enter last payday before start of sickness enter last payday before offset enter total amount paid between paydays answer no should ask which days of the week they work.  PASS (0.00s)
test_: Already getting maternity allowance Not getting maternity allowance employee told employer within time limit employee works regular days answering first sick day answering last sick day last sick day is 3 days or more after first answer yes, paid at least 8 weeks answer weekly enter last payday before start of sickness enter last payday before offset enter total amount paid between paydays answer yes enter previous sickness start date answer 6 days answer weekdays should take you to result A6.  PASS (0.02s)
test_: Already getting maternity allowance Not getting maternity allowance employee told employer within time limit employee works regular days answering first sick day answering last sick day last sick day is 3 days or more after first answer yes, paid at least 8 weeks answer weekly enter last payday before start of sickness enter last payday before offset enter total amount paid between paydays answer yes enter previous sickness start date answer 6 days should ask which days of the week do they work.  PASS (0.01s)
test_: Already getting maternity allowance Not getting maternity allowance employee told employer within time limit employee works regular days answering first sick day answering last sick day last sick day is 3 days or more after first answer yes, paid at least 8 weeks answer weekly enter last payday before start of sickness enter last payday before offset enter total amount paid between paydays answer yes enter previous sickness start date should ask how many days sick the employee had in this previous period.  PASS (0.00s)
test_: Already getting maternity allowance Not getting maternity allowance employee told employer within time limit employee works regular days answering first sick day answering last sick day last sick day is 3 days or more after first answer yes, paid at least 8 weeks answer weekly enter last payday before start of sickness enter last payday before offset enter total amount paid between paydays answer yes should ask for start date of linked period of sickness.  PASS (0.00s)
test_: Already getting maternity allowance Not getting maternity allowance employee told employer within time limit employee works regular days answering first sick day answering last sick day last sick day is 3 days or more after first answer yes, paid at least 8 weeks answer weekly enter last payday before start of sickness enter last payday before offset enter total amount paid between paydays should ask about PIW.  PASS (0.00s)
test_: Already getting maternity allowance Not getting maternity allowance employee told employer within time limit employee works regular days answering first sick day answering last sick day last sick day is 3 days or more after first answer yes, paid at least 8 weeks answer weekly enter last payday before start of sickness enter last payday before offset should ask for total amount paid.  PASS (0.00s)
test_: Already getting maternity allowance Not getting maternity allowance employee told employer within time limit employee works regular days answering first sick day answering last sick day last sick day is 3 days or more after first answer yes, paid at least 8 weeks answer weekly enter last payday before start of sickness should ask for last normal payday before payday offset.  PASS (0.00s)
test_: Already getting maternity allowance Not getting maternity allowance employee told employer within time limit employee works regular days answering first sick day answering last sick day last sick day is 3 days or more after first answer yes, paid at least 8 weeks answer weekly should ask for last payday before start sick date.  PASS (0.00s)
test_: Already getting maternity allowance Not getting maternity allowance employee told employer within time limit employee works regular days answering first sick day answering last sick day last sick day is 3 days or more after first answer yes, paid at least 8 weeks should ask how often you pay employees.  PASS (0.00s)
test_: Already getting maternity allowance Not getting maternity allowance employee told employer within time limit employee works regular days answering first sick day answering last sick day last sick day is 3 days or more after first should ask had you paid employee at least 8 weeks.  PASS (0.00s)
test_: Already getting maternity allowance Not getting maternity allowance employee told employer within time limit employee works regular days answering first sick day answering last sick day last sick day is 3 days or more after first should store last sick day.  PASS (0.00s)
test_: Already getting maternity allowance Not getting maternity allowance employee told employer within time limit employee works regular days answering first sick day answering last sick day last sick day is less than 3 days after first should take you to result A2.  PASS (0.00s)
test_: Already getting maternity allowance Not getting maternity allowance employee told employer within time limit employee works regular days answering first sick day should store response and move to Q5.  PASS (0.00s)
test_: Already getting maternity allowance Not getting maternity allowance employee told employer within time limit employee works regular days should take them to Q4.  PASS (0.00s)
test_: Already getting maternity allowance Not getting maternity allowance employee told employer within time limit should take you to Q3.  PASS (0.00s)
test_: Already getting maternity allowance Not getting maternity allowance should set adoption warning state variable.  PASS (0.00s)
test_: Already getting maternity allowance Not getting maternity allowance should take you to Q2.  PASS (0.00s)
test_: average weekly earnings is less than the LEL on sick start date should take you to result A5 as awe < LEL (as of 10/06/2013).  PASS (0.00s)
test_: no SSP payable as already had maximum should take you to result A8 as already claimed > 28 weeks (max amount).  PASS (0.01s)
test_: no SSP payable as sickness period is < 4 days should take you to result A7 - must be sick for at least 4 days in a row.  PASS (0.00s)
test_: tabular output for final SSP calculation should have consistent rates for all weekly rates that are produced.  PASS (0.09s)
test_: tabular output for final SSP calculation should have the adjusted rates in place for the week crossing through 6th April.  PASS (0.28s)
test_: tabular output for final SSP calculation should show formatted weekly payment amounts with adjusted 3 days start amount for additional SPP.  PASS (0.12s)
test_: tabular output for final SSP calculation should show formatted weekly payment amounts with adjusted 3 days start amount for ordinary SPP.  PASS (0.11s)

```

After:
```
CalculateStatutorySickPayTest
  test_: Getting additional statutory paternity pay employee didn't tell employer within time limit should go to entitled_to_sick_pay outcome.  PASS (0.22s)
  test_: Getting additional statutory paternity pay employee told employer within time limit employee works different days of the week should go to result A4.  PASS (0.00s)
  test_: Getting additional statutory paternity pay employee told employer within time limit employee works regular days should lead to entitled_to_sick_pay if worker got sick before being employed for 8 weeks and had linked sickness.  PASS (0.05s)
  test_: Getting additional statutory paternity pay employee told employer within time limit employee works regular days should lead to entitled_to_sick_pay if worker got sick before being employed for 8 weeks and had no linked sickness.  PASS (0.04s)
  test_: Getting additional statutory paternity pay employee told employer within time limit employee works regular days should lead to entitled_to_sick_pay if worker got sick before payday and had linked sickness.  PASS (0.05s)
  test_: Getting additional statutory paternity pay employee told employer within time limit employee works regular days should lead to entitled_to_sick_pay if worker got sick before payday and had no linked sickness.  PASS (0.03s)
  test_: Getting additional statutory paternity pay employee told employer within time limit employee works regular days should lead to entitled_to_sick_pay outcome when there is a linked sickness.  PASS (0.07s)
  test_: Getting additional statutory paternity pay employee told employer within time limit employee works regular days should lead to entitled_to_sick_pay without first days when there is no linked sickness.  PASS (0.20s)
  test_: Getting additional statutory paternity pay employee told employer within time limit employee works regular days should require to be sick more than 4 days to get sick pay.  PASS (0.01s)
  test_: Getting additional statutory paternity pay employee told employer within time limit should take you to Q3.  PASS (0.00s)
  test_: Getting additional statutory paternity pay should set adoption warning state variable.  PASS (0.00s)
  test_: Getting additional statutory paternity pay should take you to Q2.  PASS (0.00s)
  test_: Getting maternity allowance should go to result A1.      PASS (0.00s)
  test_: Getting statutory maternity pay should go to result A1.  PASS (0.00s)
  test_: Not getting maternity allowance should set adoption warning state variable.  PASS (0.00s)
  test_: Not getting maternity allowance should take you to Q2.   PASS (0.00s)
  test_: average weekly earnings is less than the LEL on sick start date should take you to result A5 as awe < LEL (as of 2013-06-10).  PASS (0.00s)
  test_: no SSP payable as already had maximum should take you to result A8 as already claimed > 28 weeks (max amount).  PASS (0.01s)
  test_: no SSP payable as sickness period is < 4 days should take you to result A7 - must be sick for at least 4 days in a row.  PASS (0.00s)
  test_: tabular output for final SSP calculation should have consistent rates for all weekly rates that are produced.  PASS (0.17s)
  test_: tabular output for final SSP calculation should have the adjusted rates in place for the week crossing through 6th April.  PASS (0.11s)
  test_: tabular output for final SSP calculation should show formatted weekly payment amounts with adjusted 3 days start amount for additional SPP.  PASS (0.23s)
  test_: tabular output for final SSP calculation should show formatted weekly payment amounts with adjusted 3 days start amount for ordinary SPP.  PASS (0.22s)
```

**Comment by chrisroos** [comment]
What's the plan for this pull request?

**Comment by BenJanecke** [comment]
@chrisroos This branch was rebased into #1634 so I think we can close this PR

**Comment by tadast** [comment]
Yes, we can move the discussion about flat tests to https://github.com/alphagov/smart-answers/pull/1634 as there is a concrete situation to help us reason about it (https://github.com/alphagov/smart-answers/commit/96a9d2eea94e152afed830c717a123ced6751bd0)
